### DISCUSSION TITLE
Did the API shift how it handles attachment objects for deleted messages?

### DISCUSSION THREAD

**Discussion by ghost** [Rapptz/discord.py#9858] [Q&A] [answered]
Used to be I could just do something along the lines of this to upload a copy of attachments to a separate channel after the message was deleted for logging purposes (IE: Images, Videos, etc)

```python
attachments = []
attachment_links = []
if len(message.attachments) > 0:
    for attachment in message.attachments:
        attachments.append(await attachment.to_file())
        attachment_links.append(attachment.url)

attachment_urls = '\n'.join(attachment_links)
embed.add_field(name='Attachments', value=attachment_urls, inline=False)
await channel.send(embed=embed)
await channel.send(files=attachments)
```

Which used to work without fail 100% of the time- (I assume because attachment links had a time lag before they expired after message deletion)
But now the results are either extremely inconsistent (working only 1/20 or so times), or just end in outright failure.
I can only really assume this is due to a change in how Discord themselves handles attachments after the message they're tied to is deleted.

This thought is further reinforced by looking at the error raised in question

```python
2024-06-06 14:59:06 ERROR    discord.client Ignoring exception in on_message_delete
Traceback (most recent call last):
  File "/home/peppermint/vpy/lib/python3.12/site-packages/discord/client.py", line 441, in _run_event
    await coro(*args, **kwargs)
  File "/home/pep/Desktop/Development/DiscordBots/Hybrid/Spoopy/test.py", line XX, in on_message_delete
    attachments.append(await attachment.to_file())
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/pep/vpy/lib/python3.12/site-packages/discord/message.py", line 325, in read
    data = await self._http.get_from_cdn(url)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/pep/vpy/lib/python3.12/site-packages/discord/http.py", line 769, in get_from_cdn
    raise NotFound(resp, 'asset not found')
discord.errors.NotFound: 404 Not Found (error code: 0): asset not found
```

Just looking for confirmation on whether or not the API changed, or I need to update my code- or I'm simply going insane.

**Comment by Rapptz** [comment, marked as answer]
This has never worked. At best, proxy_urls would be cached for a bit longer than the parent attachment URL. This has been [mentioned in the documentation](https://discordpy.readthedocs.io/en/stable/api.html#discord.Attachment.proxy_url) for years as far as I know.

**Reply by ghost** [reply]
Strange- I did tests on it and it worked fine for *months*. And I'm only just now encountering issues.
As I'm relatively new to the whole Discord development scene, and am, lets face it, not as well versed as I could be- I got confused as to the inconsistency

**Reply by BuIlDaLiBlE** [reply]
Seems like something has changed on Discord's side, which is unfortunate.

**Reply by ghost** [reply]
After more testing- figured out that (probably) isn't the case. The reason it was working just fine before was because I was testing it in a populated server.

The reason the uploads were working, was because if another user views the attachment, it apparently ups the live time of the normal CDN URL (But not the proxy. Which is weird, since proxy is supposed to be the cached version). That allowed the uploads to work really often, which I mistakenly assumed to just be the default behavior.

As for the reason I was getting the error in more recent testing- this was because I was rewriting the bot, and testing all the events in my private server (of which I'm the only member). And because of there being no third-party to view the attachment, the link gets purged immediately.
